refactor: 사용자 정보 파싱 완료

eaststar113 opened the pull request:
### 이슈
ERD참고하여 사용자 정보 파싱 진행

***

### 수정사항 및 반영

사용자 정보 총 7개 파싱하였고 역할은 admin과 student 두가지 존재합니다. 피시너들의 정보는 포함이 안되며 현재 블랙홀로 빠져나간 인원, 보컬 의 정보들이 포함되어 있습니다. 레벨은 double형태로 받았으나 소수점의 존재로 길이는 제각각 다르게 파싱되어있습니다.(ex) 1.6000000002, 3.2, 0.0 )

@joejaeyoung @joonwan @ghkgus

eaststar113 committed:
refactor: 사용자 정보 파싱 완료

joonwan reviewed (commented):
고생하셨습니다!

joonwan commented on src/test/java/com/ggs/STAT_TurboFetch/client/FtClientImplTest.java:
```diff
@@ -0,0 +1,31 @@
+        assertEquals(5, userPersonalDto.getWallet());
+        assertEquals(6, userPersonalDto.getCorrection_point());
+    }
+}
```
실패 테케도 작성해주시면 너무 좋을것 같 아 요!

eaststar113 replied:
예외 상황 확인하는 테스트 케이스 추가하였습니다

ghkgus commented on src/main/java/com/ggs/STAT_TurboFetch/client/FtClientImpl.java:
```diff
@@ -74,9 +77,66 @@ public List<String> getAllUsers(String code) {
+        return (Double) data.get("level");
+    }
+
+    public static UserPersonalDto extractUserField(HashMap<String, Object> user){
```
public이 private보다 위에 있으면 좋을 것 같습니다.

eaststar113 replied:
수정하겠습니다

joonwan reviewed (changes requested):
커멘트 확인해주세요!!!

joejaeyoung commented on src/main/java/com/ggs/STAT_TurboFetch/api/controller/UserController.java:
```diff
@@ -19,8 +20,8 @@ public class UserController {
     @GetMapping("/users")
-    public ApiResponse<List<String>> getAllUsers(
-            @RequestParam(required = false) String code ) {
+    public ApiResponse<List<UserDto>> getAllUsers(
```
이님, DTO를 돌려쓰시던데.. 이건..
1. 순환 참조 문제
2. 단방향 의존성 원칙 위배
3. 관심사의 분리 훼손
4. 유지보수성 저하
5. 재사용성 감소
6. 테스트 용이성 감소

생긴다 배웠는데 어떻게 생각하시나요?

eaststar113 replied:
계층 간의 관계성에 대해 생각지 못하고 단순 편의를 위해 dto를 중복해서 사용했는데 적어주신 이유들을 바탕으로 입력받을 때 사용하는 dto랑 출력할 때 사용하는 dto를 분리해서 사용해야된다는 것을 덕분에 알게 되었습니다. 단일 책임 원칙을 위배하게 되버렸네요; 후에 수정하겠습니다

joejaeyoung commented on src/main/java/com/ggs/STAT_TurboFetch/client/FtClientImpl.java:
```diff
@@ -74,9 +77,66 @@ public List<String> getAllUsers(String code) {
+        return parsingResponseList;
+    }
+
+    private static UserDto setUserDtoField(Double level, UserPersonalDto userPersonalDto, String largeImage){
```
@Setter 사용대신 메소드를 새로 만드신 이유가 궁금합니다.

eaststar113 replied:
@Setter를 직접 사용하지는 않았지만, @Data 어노테이션을 사용했기 때문에 내부적으로 @Setter가 포함되어 있습니다. 
setter사용을 지양하는게 장기적으로 좋은 방법(이유 - 캡슐화, 테스트 및 디버깅 용이, 설계 원칙(SOLID) 준수, 유지보수성 향상 등)이기에 @Getter와 @RequiredArgsConstructor를 @Data 대신 사용하거나 build패턴을 사용하는 것도 좋은 방법인듯 싶네요

joejaeyoung replied:
감사합니다.

eaststar113 replied:
빌드 패턴 적용하여 setter사용 지양하였습니다

joejaeyoung commented on src/test/java/com/ggs/STAT_TurboFetch/client/FtClientImplTest.java:
테스트 코드가 너무 짧습니다..

eaststar113 replied:
테스트케이스 하나 추가했습니다

joejaeyoung reviewed (changes requested):
고생하셨습니다.
확인 부탁드립니다.

ghkgus reviewed (commented):
고생하셨습니다.

eaststar113 committed:
refactor: 수정사항 반영
- 삼항 연산자 사용 이유는 erd를 참고했을 때 자료형이 integer가 아닌 int이기에 사용
- setter사용을 지양하기 위해 빌드 패턴 사용
- 예외 발생 테스트 케이스 작성

eaststar113 committed:
fix: test 계정 삭제 및 삼항연산자를 if문으로 대체

joonwan reviewed (commented):
고생하셨습니다.

joonwan reviewed (commented):
고생하셨습니다.

eaststar113 committed:
fix: test account 삭제 및 메서드명 변경
